include/LcdFloatSpinBox.h

```diff
@@ -0,0 +1,97 @@
+/*
+ * LcdFloatSpinBox.h - class LcdFloatSpinBox (LcdSpinBox for floats)
+ *
+ * Copyright (c) 2005-2014 Tobias Doerffel <tobydox/at/users.sourceforge.net>
+ * Copyright (c) 2020 Martin Pavelek <he29.HS/at/gmail.com>
+ *
+ * This file is part of LMMS - https://lmms.io
+ *
+ * This program is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU General Public
+ * License as published by the Free Software Foundation; either
+ * version 2 of the License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public
+ * License along with this program (see COPYING); if not, write to the
+ * Free Software Foundation, Inc., 51 Franklin Street, Fifth Floor,
+ * Boston, MA 02110-1301 USA.
+ *
+ */
+
+
+#ifndef LCD_FLOATSPINBOX_H
+#define LCD_FLOATSPINBOX_H
+
+#include <QString>
+
+#include "LcdWidget.h"
+#include "AutomatableModelView.h"
+
+
+class LMMS_EXPORT LcdFloatSpinBox : public QWidget, public FloatModelView
+{
+	Q_OBJECT
+public:
+	LcdFloatSpinBox(int numWhole, int numFrac, QWidget* parent, const QString& name = QString());
+	LcdFloatSpinBox(int numWhole, int numFrac, const QString& style, QWidget* parent, const QString& name = QString());
+	virtual ~LcdFloatSpinBox() = default;
+
+	void modelChanged() override
+	{
+		ModelView::modelChanged();
+		update();
+	}
+
+	/*! Sets an offset which is always added to value of model so we can
+	    display values in a user-friendly way if they internally start at 0 */
+	void setDisplayOffset(int offset)
+	{
+		m_displayOffset = offset;
+	}
+
+	/*! \brief Returns internal offset for displaying values */
+	int displayOffset() const
+	{
+		return m_displayOffset;
+	}
+
+	void setLabel(const QString &label) { m_label = label; }
+
+public slots:
+	virtual void update();
```

Comment:
Can this have a name other than `update`? The base class function isn't virtual, so this function shadows it, not overrides it. I don't know how this interacts with the signal-slot mechanism, but it may not behave as the implementation/callers intended. Also, it needn't be marked virtual, as it doesn't look like it's intended to be overridden.

Reply:
I did not even think of any of this, I just copied it from `LcdSpinBox` and assumed it will work the same. I would guess it still needs to be called `update()` since that is what Qt calls when it plans to redraw a widget?

Reply:
Since `QWidget::update()` is non-virtual, the name of this function doesn't matter.

Reply:
I know it doesn't matter, but I think the name is misleading, shadowing `QWidget::update` with something different. Also, slot lookups are done using strings, not vtables, so it's still not clear how this interacts with the signal-slot mechanism. I'll investigate it.

Reply:
To clarify, I was answering to @he29-net's question.

Reply:
Ah, that makes sense 😄. It's late and I'm not reading properly. I still want to investigate how slot name shadowing behaves though.

Reply:
If I remember correctly, Qt looks at the meta object of the target class for string-based(`SIGNAL()`, `SLOT()` macros) connections. You can find something like `src/lmmsobjs_autogen/DGKZTCOIDE/moc_LcdSpinBox.cpp` in your build directory which contains signal/slot information.

Reply:
I just tried to rename it and the display GUI stopped updating. So Qt (or LMMS, not sure who calls the global `update()` on all widgets) definitely expects the slot to be named `update()`.

Reply:
I tested this, see the code at https://gist.github.com/DomClark/fc976514c6e1b083d9bce0ab777eefad. It appears slots have virtual behaviour when connected using `SIGNAL`/`SLOT` macros, and non-virtual behaviour when using member function pointer connections. This inconsistency, along with the fact that it doesn't work with the modern connection style, suggests to me that it's not a good idea to override slots by name. I'm fine with leaving it as is for now, since the existing `LcdSpinBox` code does the same thing, but I think we should clean this up in the future.

Reply:
That inconsistency certainly looks like an invitation for some confusing bugs, interesting..

I wonder what is the intended workflow then? I would expect that a derived widget implements its own `update()`, so that it can react to any changes specific to the modified version. And as a part of its `update()` it would call `update()` of the base class to finish the rest. So the non-virtual behavior should be almost always unwanted, right?

But if `update()`s are called by Qt on some occasions, wouldn't choosing a different "update" slot name on a derived class mean that it simply won't be called when needed? Or do we only call / signal `update()`s explicitly from our code?

Reply:
`update()` isn't supposed actually update anything itself. Its purpose is to schedule a repaint of the widget in the near future, once control returns to the event loop. If virtual behaviour were intended, it would be a virtual function.

It looks like the code there is to set the LCD display to the current value of the model, in which case it should run in response to changes in the model. The appropriate places for this are in `LcdFloatSpinBox::modelChanged` (called when a new model is viewed by the spin box) and in response to the `Model::dataChanged` signal (emitted when the model's value changes).

Reply:
In addition to @DomClark's comment, `ModelView::doConnections` connects `Model::dataChanged` and `Model::propertiesChanged` to `SLOT(update())` of `ModelView::m_widget`. This means at least one of them should be satisfied:
- `paintEvent()` of the model reads the value of the model,
- The view monitors changes in its model
- The view implements its own `update()`

`LcdFloatSpinBox`(and `LcdSpinBox`, too) is using the third approach. If this approach is not recommended, we should move to other approaches.
However, this looks out of scope of this PR.